test/core/base/cache.js

```diff
@@ -1,32 +1,160 @@
-describe('axe._cache', function() {
+describe('axe._cache', function () {
   'use strict';
 
-  it('should set items from the cache without error', function() {
-    function fn() {
-      axe._cache.set('foo', 'bar');
-    }
-    assert.doesNotThrow(fn);
-  });
+  describe('set', function () {
+    it('should set items without error', function () {
+      function fn() {
+        axe._cache.set('foo', 'bar');
+      }
+      assert.doesNotThrow(fn);
+    });
+
+    it('should set `undefined` without error', function () {
+      function fn() {
+        axe._cache.set('foo', undefined);
+        assert.isUndefined(axe._cache.get('foo'));
+      }
+      assert.doesNotThrow(fn);
+    });
+
+    it('should throw TypeError for non-string keys', function () {
+      assert.throws(function () {
+        axe._cache.set(1, 'bar');
+      }, TypeError);
+      assert.throws(function () {
+        axe._cache.set({}, 'bar');
+      }, TypeError);
+      assert.throws(function () {
+        axe._cache.set(null, 'bar');
+      }, TypeError);
+      assert.throws(function () {
+        axe._cache.set(function () {
+          return;
+        }, 'bar');
+      }, TypeError);
+    });
 
-  it('should not throw for non-string keys', function() {
-    function fn() {
-      axe._cache.set(1, 'bar');
-      axe._cache.set({}, 'bar');
-      axe._cache.set(null, 'bar');
-    }
-    assert.doesNotThrow(fn);
+    it('should throw TypeError for empty string keys', function () {
+      assert.throws(function () {
+        axe._cache.set('', 'bar');
+      }, TypeError);
+    });
   });
 
-  it('should get an item from the cache', function() {
-    axe._cache.set('foo', 'bar');
-    var value = axe._cache.get('foo');
-    assert.equal(value, 'bar');
+  describe('get', function () {
+    it('should get an item from the cache', function () {
+      axe._cache.set('foo', 'bar');
+      var value = axe._cache.get('foo');
+      assert.equal(value, 'bar');
+    });
+
+    it('should return `undefined` for key lookup miss', function () {
+      assert.isUndefined(axe._cache.get('foo'));
+    });
+
+    describe('with default value', function () {
+      it('should set and return primitive types', function () {
+        assert.equal(
+          axe._cache.get('integer', function () {
+            return 1;
+          }),
+          1
+        );
+        assert.equal(
+          axe._cache.get('float', function () {
+            return 1.1;
+          }),
+          1.1
+        );
+        assert.equal(
+          axe._cache.get('string', function () {
+            return 'foo';
+          }),
+          'foo'
+        );
+        assert.isTrue(
+          axe._cache.get('boolean', function () {
+            return true;
+          })
+        );
+        assert.isNull(
+          axe._cache.get('null', function () {
+            return null;
+          })
+        );
+        assert.deepEqual(
```

Comment:
We should check equality, not just deepEqual. We want it to be the same object that's returned.